assembly/src/parsers/ast/ast.rs

```diff
@@ -0,0 +1,256 @@
+use vm_core::utils::collections::BTreeMap;
+
+use vm_core::Felt;
+
+// AST
+// ================================================================================================
+
+// This holds the list of instructions supported in a Miden program.
+// This instruction list is used to hold reference to the instruction, and future be
+// used for MAST generation.
+pub enum Instruction {
+    Assert,
+    AssertEq,
+    Assertz,
+    Add,
+    AddImm(Felt),
+    Sub,
+    SubImm(Felt),
+    Mul,
+    MulImm(Felt),
+    Div,
+    DivImm(Felt),
+    Neg,
+    Inv,
+    Pow2Checked,
+    Pow2Unchecked,
+    Not,
+    And,
+    Or,
+    Xor,
+    Eq,
+    EqImm(Felt),
+    Neq,
+    NeqImm(Felt),
+    Eqw,
+    Lt,
+    Lte,
+    Gt,
+    Gte,
+
+    // ----- u32 manipulation ---------------------------------------------------------------
+
+    U32Test,
+    U32TestW,
+    U32Assert,
+    U32AssertW,
+    U32Split,
+    U32Cast,
+    U32CheckedAdd,
+    U32CheckedAddImm(Felt),
+    U32WrappingAdd,
+    U32WrappingAddImm(Felt),
+    U32OverflowingAdd,
+    U32OverflowingAddImm(Felt),
+    U32OverflowingAdd3,
+    U32WrappingAdd3,
+    U32CheckedSub,
+    U32CheckedSubImm(Felt),
+    U32WrappingSub,
+    U32WrappingSubImm(Felt),
+    U32OverflowingSub,
+    U32OverflowingSubImm(Felt),
+    U32CheckedMul,
+    U32CheckedMulImm(Felt),
+    U32WrappingMul,
+    U32WrappingMulImm(Felt),
+    U32OverflowingMul,
+    U32OverflowingMulImm(Felt),
+    U32OverflowingMadd,
+    U32WrappingMadd,
+    U32CheckedDiv,
+    U32CheckedDivImm(Felt),
+    U32UncheckedDiv,
+    U32UncheckedDivImm(Felt),
+    U32CheckedMod,
+    U32CheckedModImm(Felt),
+    U32CheckedDivMod,
+    U32CheckedDivModImm(Felt),
+    U32UncheckedDivMod,
+    U32UncheckedDivModImm(Felt),
+    U32CheckedAnd,
+    U32CheckedOr,
+    U32CheckedXor,
+    U32CheckedNot,
+    U32CheckedShr,
+    U32CheckedShrImm(Felt),
+    U32UncheckedShr,
+    U32UncheckedShrImm(Felt),
```

Comment:
For shifts/rotations we can set the data type of the immediate values to be `u8`. Or maybe a better option is to create a newtype which would guarantee that values are in the appropriate range.

Reply:
I'll use u8 for now